web/api/tests/web_api.c

```diff
@@ -126,43 +211,234 @@ static void build_request(struct web_buffer *wb, const char *url, bool use_cr, s
     buffer_strcat(wb, "\n");
 }
 
-static struct web_client *pre_test_setup()
+/* Note: this is not a CMocka group_test_setup/teardown pair. This is performed per-test.
+*/
+static struct web_client *setup_fresh_web_client()
 {
-    localhost = malloc(sizeof(RRDHOST));
     struct web_client *w = (struct web_client *)malloc(sizeof(struct web_client));
+    memset(w, 0, sizeof(struct web_client));
     w->response.data = buffer_create(NETDATA_WEB_RESPONSE_INITIAL_SIZE);
     w->response.header = buffer_create(NETDATA_WEB_RESPONSE_HEADER_SIZE);
     w->response.header_output = buffer_create(NETDATA_WEB_RESPONSE_HEADER_SIZE);
     strcpy(w->origin, "*"); // Simulate web_client_create_on_fd()
     w->cookie1[0] = 0;      // Simulate web_client_create_on_fd()
     w->cookie2[0] = 0;      // Simulate web_client_create_on_fd()
+    w->acl = 0x1f;          // Everything on
     return w;
 }
 
-static void post_test_cleanup(struct web_client *w)
+static void destroy_web_client(struct web_client *w)
 {
     buffer_free(w->response.data);
     buffer_free(w->response.header);
     buffer_free(w->response.header_output);
     free(w);
-    free(localhost);
 }
 
+// ---------------------------------- Parameterized test-families -----------------------------------------------------
+// There is no way to pass a parameter block into the setup fixture, we would have to patch CMocka and maintain it
```

Comment:
[cmocka_unit_test_prestate_setup_teardown(f, setup, teardown, state)](https://api.cmocka.org/group__cmocka__exec.html#ga07ea44503d0f6c84c460b4a878cf31d5)

Reply:
The state is passed into the tests if nothing is returned from the setup, it is not passed into the setup. (L3441) in https://git.cryptomilk.org/projects/cmocka.git/tree/src/cmocka.c

You are correct about what the docs say, but it did not do that when I tested it and the source of CMocka seems to disagree with their documentation.